### PR TITLE
GigaSpeech RNN-T experiments

### PR DESCRIPTION
**Pull request by wgb14**

### CODE REVIEW

**Comment by wgb14** [comment]
Post Dan's Suggestions here:

> Guanbo, if you want to run an RNN-T recipe, please now use the setup in pruned_transducer_stateless2 from librispeech. This converges much faster than the old setup. The only options you might want to change are: --lr-epochs (reduce from 6 to some number less than about half the number of epochs you plan to train, e.g. 2 or 3). You might want to add --use-fp16=True to use half-precision to speed up training... and you can adjust --max-duration accordingly. --use-fp16=True allows, on our GPUs, to increase max-duration from 300 to 550. (However it will converge slower near the start of training if the max-duration is too high, so this is a tradeoff).

**Comment by wgb14** [comment]
However, each time I set a larger `--max-duration`, either in Conformer-CTC or RNN-T scripts, I got this error:

`--max-duration 200`: (during `scan_pessimistic_batches_for_oom`)
```log
Traceback (most recent call last):
  File "./pruned_transducer_stateless2/train.py", line 977, in <module>
    main()
  File "./pruned_transducer_stateless2/train.py", line 968, in main
    mp.spawn(run, args=(world_size, args), nprocs=world_size, join=True)
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 230, in spawn
    return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 188, in start_processes
    while not context.join():
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 150, in join
    raise ProcessRaisedException(msg, error_index, failed_process.pid)
torch.multiprocessing.spawn.ProcessRaisedException: 

-- Process 3 terminated with the following error:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 59, in _wrap
    fn(i, *args)
  File "/userhome/user/guanbo/icefall_rnnt/egs/gigaspeech/ASR/pruned_transducer_stateless2/train.py", line 863, in run
    params=params,
  File "/userhome/user/guanbo/icefall_rnnt/egs/gigaspeech/ASR/pruned_transducer_stateless2/train.py", line 944, in scan_pessimistic_batches_for_oom
    loss.backward()
  File "/opt/conda/lib/python3.7/site-packages/torch/_tensor.py", line 307, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/__init__.py", line 156, in backward
    allow_unreachable=True, accumulate_grad=True)  # allow_unreachable flag
RuntimeError: CUDA error: invalid configuration argument
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
```

`--max-duration 150`: (during `train_one_epoch`)
```log
/opt/conda/lib/python3.7/multiprocessing/semaphore_tracker.py:144: UserWarning: semaphore_tracker: There appear to be 11 leaked semaphores to clean up at shutdown
  len(cache))
/opt/conda/lib/python3.7/multiprocessing/semaphore_tracker.py:144: UserWarning: semaphore_tracker: There appear to be 11 leaked semaphores to clean up at shutdown
  len(cache))
Traceback (most recent call last):
  File "./pruned_transducer_stateless2/train.py", line 977, in <module>
    main()
  File "./pruned_transducer_stateless2/train.py", line 968, in main
    mp.spawn(run, args=(world_size, args), nprocs=world_size, join=True)
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 230, in spawn
    return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 188, in start_processes
    while not context.join():
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 150, in join
    raise ProcessRaisedException(msg, error_index, failed_process.pid)
torch.multiprocessing.spawn.ProcessRaisedException: 

-- Process 3 terminated with the following error:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 59, in _wrap
    fn(i, *args)
  File "/userhome/user/guanbo/icefall_rnnt/egs/gigaspeech/ASR/pruned_transducer_stateless2/train.py", line 892, in run
    rank=rank,
  File "/userhome/user/guanbo/icefall_rnnt/egs/gigaspeech/ASR/pruned_transducer_stateless2/train.py", line 686, in train_one_epoch
    scaler.scale(loss).backward()
  File "/opt/conda/lib/python3.7/site-packages/torch/_tensor.py", line 307, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/__init__.py", line 156, in backward
    allow_unreachable=True, accumulate_grad=True)  # allow_unreachable flag
RuntimeError: CUDA error: invalid configuration argument
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
```
It is said this `CUDA error: invalid configuration argument` is another type of CUDA OOM. The GPU cards I'm using has 32GB of memory. In my experiments, i can only set `--max-duration 120` for `bpe_500` and  `--max-duration 150` for `bpe_5000`, both in Conformer-CTC and RNN-T. I cannot benefit from `use-fp16`.

**Comment by csukuangfj** [comment]
I find that you have removed the following block from `train.py`:
https://github.com/k2-fsa/icefall/blob/fce7f3cd9a486405ee008bcbe4999264f27774a3/egs/librispeech/ASR/pruned_transducer_stateless2/train.py#L849-L860

Could you restore it? You may need to adjust the max duration threshold for GigaSpeech.
From my experience, 20 is also a good choice for the GigaSpeech dataset.

**Comment by wgb14** [comment]
> I find that you have removed the following block from `train.py`:
> 
> https://github.com/k2-fsa/icefall/blob/fce7f3cd9a486405ee008bcbe4999264f27774a3/egs/librispeech/ASR/pruned_transducer_stateless2/train.py#L849-L860
> 
> Could you restore it? You may need to adjust the max duration threshold for GigaSpeech. From my experience, 20 is also a good choice for the GigaSpeech dataset.

Gigaspeech has already filtered out segments longer than 20 s, but kept those shorter than 1 s. I can try this though, since we did speed perturbation, and some utterances could be 22 s long.
But I wonder if it is 20 s to 22 s that caused `max-duration` dropped from 550 to 120.

**Comment by csukuangfj** [comment]
I am also using GigaSpeech dataset for training in https://github.com/k2-fsa/icefall/pull/312

It works very well with `--max-duration 300` after applying `remove_short_and_long_utt()`

**Comment by csukuangfj** [comment]
>  I can try this though, since we did speed perturbation, and some utterances could be 22 s long.

Yes, I also noticed that. But such utterances account for only less than 0.1% of the data. It is safe to remove them, I think.

**Comment by danpovey** [comment]
Our pruned RNN-T is quite memory efficient for longer utterances; I'd be more worried about shorter ones, since constant factors can sometimes be important.  But we need to be operating from definite knowledge.  Please add try/except to catch the error and print out the failing batch's details (I thought we had this code at one point); and also, to catch things earlier possibly, do
export CUDA_LAUNCH_BLOCKING=1
export K2_SYNC_KERNELS=1

**Comment by danpovey** [comment]
... BTW, Desh Raj @desh2608 has reported a similar problems at CLSP.  I'm not sure the cause, but we need to try hard to debug this thoroughly.  I'd also maybe try installing nsys and running it under "nsys profile".. that may give us additional info.

**Comment by danpovey** [comment]
...  another possibility, which is a bit ugly but would work if the problem is short utterances, would be to enforce a maximum number of utterances per batch, i.e. a maximum batch size, e.g. 512, in the main training loop, by just discarding any extra elements.

**Comment by pzelasko** [comment]
> ... another possibility, which is a bit ugly but would work if the problem is short utterances, would be to enforce a maximum number of utterances per batch, i.e. a maximum batch size, e.g. 512, in the main training loop, by just discarding any extra elements.

Just add `max_cuts=512` in the sampler.

**Comment by wgb14** [comment]
> > ... another possibility, which is a bit ugly but would work if the problem is short utterances, would be to enforce a maximum number of utterances per batch, i.e. a maximum batch size, e.g. 512, in the main training loop, by just discarding any extra elements.
> 
> Just add `max_cuts=512` in the sampler.

Do we support `max_cuts` in `DynamicBucketingSampler`? I didn't see that in 
https://github.com/lhotse-speech/lhotse/blob/b3b96a1d64fa3bd97c9c9bf32ef0e8b4806b87ef/lhotse/dataset/sampling/dynamic_bucketing.py#L25

**Comment by pzelasko** [comment]
Oh right, dynamic bucketing is the only one that does not support it yet... it shouldn't be too complicated but I'd like to test it properly, if you need it feel free to contribute.

**Comment by pzelasko** [comment]
OK I found a moment to add it after all, try this PR https://github.com/lhotse-speech/lhotse/pull/681

**Comment by wgb14** [comment]
> Our pruned RNN-T is quite memory efficient for longer utterances; I'd be more worried about shorter ones, since constant factors can sometimes be important. But we need to be operating from definite knowledge. Please add try/except to catch the error and print out the failing batch's details (I thought we had this code at one point); and also, to catch things earlier possibly, do export CUDA_LAUNCH_BLOCKING=1 export K2_SYNC_KERNELS=1

Yes, we do have this detail, but in original script it only print when the error is `"CUDA out of memory"`. By changing to print whatever error is, I got this log:
```log
2022-04-21 09:36:13,058 ERROR [train_fix.py:964] (1/8) Your GPU ran out of memory with the current max_duration setting. We recommend decreasing max_duration and trying again.
Failing criterion: largest_batch_cuts_duration (=300.0) ...
Traceback (most recent call last):
  File "./pruned_transducer_stateless2/train_fix.py", line 991, in <module>
    main()
  File "./pruned_transducer_stateless2/train_fix.py", line 982, in main
    mp.spawn(run, args=(world_size, args), nprocs=world_size, join=True)
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 230, in spawn
    return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 188, in start_processes
    while not context.join():
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 150, in join
    raise ProcessRaisedException(msg, error_index, failed_process.pid)
torch.multiprocessing.spawn.ProcessRaisedException: 

-- Process 1 terminated with the following error:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/multiprocessing/spawn.py", line 59, in _wrap
    fn(i, *args)
  File "/userhome/user/guanbo/icefall_rnnt/egs/gigaspeech/ASR/pruned_transducer_stateless2/train_fix.py", line 877, in run
    params=params,
  File "/userhome/user/guanbo/icefall_rnnt/egs/gigaspeech/ASR/pruned_transducer_stateless2/train_fix.py", line 958, in scan_pessimistic_batches_for_oom
    loss.backward()
  File "/opt/conda/lib/python3.7/site-packages/torch/_tensor.py", line 307, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/__init__.py", line 156, in backward
    allow_unreachable=True, accumulate_grad=True)  # allow_unreachable flag
RuntimeError: CUDA error: invalid configuration argument
```
So the Failing criterion is still the largest_batch_cuts_duration (=300.0), instead of "max_num_cuts"

By the way, I got this error with `--max-duration 300` even after applying `remove_short_and_long_utt()`. And 
```bash
export CUDA_LAUNCH_BLOCKING=1
export K2_SYNC_KERNELS=1
```
didn't give me any additional information.

**Comment by wgb14** [comment]
Post Tensorboard so far: https://tensorboard.dev/experiment/RkuNltQOR9aI7BGr5AHbcA/
Can anybody give some comments and suggestions? This seems not natural to me.

**Comment by csukuangfj** [comment]
So do you finally manage to train it with max duration == 300?

The pruned loss looks ok, I think.  

There sould be some checkpoints in your exp dir. You can try to decode with them and see the WER.

**Comment by csukuangfj** [comment]
As for the error,
> RuntimeError: CUDA error: invalid configuration argument

There is a similar issue at https://github.com/pytorch/pytorch/issues/48573, which has been fixed in https://github.com/pytorch/pytorch/pull/64194

Wondering if you are using some recent versions of PyTorch.

**Comment by csukuangfj** [comment]
> Can anybody give some comments and suggestions? This seems not natural to me.

The following is the tensorboard log for https://github.com/k2-fsa/icefall/pull/312, which uses 10% of the speed perturbed GigaSpeech data (i.e., 3k hours) for training.

https://tensorboard.dev/experiment/lVZFGwjKS9iMYHzrsAE2cw/#scalars&_smoothingWeight=0

And the following is the WER of that PR for the GigaSpeech dev/test datasets:
```
(py38) kuangfangjun:modified_beam_search$ grep -r -n --color "best for test" log-* | sort -n -k2 | head  -n5
log-decode-epoch-34-avg-11-beam-4-2022-04-20-00-50-01:125:beam_size_4   12.33   best for test
log-decode-epoch-34-avg-16-beam-4-2022-04-20-02-22-16:125:beam_size_4   12.33   best for test
log-decode-epoch-34-avg-17-beam-4-2022-04-20-02-38-56:125:beam_size_4   12.33   best for test
log-decode-epoch-34-avg-18-beam-4-2022-04-20-02-55-32:125:beam_size_4   12.33   best for test
log-decode-epoch-34-avg-20-beam-4-2022-04-20-03-28-57:125:beam_size_4   12.33   best for test
(py38) kuangfangjun:modified_beam_search$ grep -r -n --color "best for dev" log-* | sort -n -k2 | head  -n5
log-decode-epoch-34-avg-15-beam-4-2022-04-20-02-04-44:170:beam_size_4   12.23   best for dev
log-decode-epoch-34-avg-16-beam-4-2022-04-20-02-22-16:170:beam_size_4   12.23   best for dev
log-decode-epoch-34-avg-12-beam-4-2022-04-20-01-08-26:170:beam_size_4   12.25   best for dev
log-decode-epoch-34-avg-14-beam-4-2022-04-20-01-46-06:170:beam_size_4   12.25   best for dev
log-decode-epoch-34-avg-17-beam-4-2022-04-20-02-38-56:170:beam_size_4   12.25   best for dev
```

In your tensorboard log, the pruned loss is around 0.7, which seems to be normal.  Since you are using all the training data, I believe your WERs will be lower.

**Comment by wgb14** [comment]
Thanks, Fangjun. These really help a lot.
No, I'm still training with `--max-duration 120`, same with Conformer-CTC setup. I'm using torch 1.10.0, a verion after that fix.

**Comment by wgb14** [comment]
Update results here: (only on DEV set)

||  greedy search | fast beam search | modified beam search |
|------|-------|-------|-------|
| WER | 10.76 | 10.71 | 10.65 |

The RNN-T model was trained for 20 epochs, and the best numbers are from the 20th epoch (--epoch 19 --avg 6), while in Conformer CTC model, the best numbers are from the 19th epoch (  --epoch 18 --avg 6).
Not sure if I can get better numbers if I continue to train RNN-T for 10 more epochs.

**Comment by danpovey** [comment]
Cool!

**Comment by wgb14** [comment]
Results are:

|                      |  Dev  | Test  |
|----------------------|-------|-------|
|    greedy search     | 10.59 | 10.87 |
|   fast beam search   | 10.56 | 10.80 |
| modified beam search | 10.52 | 10.62 |

This PR is ready for review

**Comment by csukuangfj** [comment]
Thanks!  Looks great to me.

Please remove WIP when you think it is ready to merge.

**Comment by wgb14** [comment]
Ready to merge now. 

By the way, we would like to add the results to gigaspeech leaderboard, what's the preferred name of this recipe/model? RNN-T or something like Pruned Stateless Transducer?

**Comment by danpovey** [comment]
Maybe pruned stateless RNN-T?
The conformer model in pruned_transducer_stateless2 also has a lot of changes from vanilla conformer but it doesn't really have a name, so I think " pruned stateless RNN-T" is OK for now.

File: egs/gigaspeech/ASR/RESULTS.md [hunk 1]
```diff
@@ -3,6 +3,13 @@
 
 #### 2022-05-12
 
+#### Conformer encoder + embedding decoder
+
+Conformer encoder + non-current decoder. The encoder a reworked
```
Comment (csukuangfj) [inline egs/gigaspeech/ASR/RESULTS.md:8]:
```suggestion
Conformer encoder + non-recurrent decoder. The encoder is a reworked 
```

**Comment by wgb14** [comment]
> Maybe pruned stateless RNN-T? The conformer model in pruned_transducer_stateless2 also has a lot of changes from vanilla conformer but it doesn't really have a name, so I think " pruned stateless RNN-T" is OK for now.

I see.  I will add some details in `RESULTS.md`

File: egs/gigaspeech/ASR/RESULTS.md (hunk 1 above)
Comment (wgb14) [reply egs/gigaspeech/ASR/RESULTS.md]: Fixed

**Comment by csukuangfj** [comment]
Ok, I am merging it.

**Comment by csukuangfj** [comment]
@wgb14

Could you also upload the decoding results, e.g., files like `log-*, errs-*, recogs-*`, to huggingface?

**Comment by csukuangfj** [comment]
I just reproduced the greedy search results locally. Here are the decoding logs:
```
2022-05-13 11:28:21,598 INFO [decode.py:478] Decoding started
2022-05-13 11:28:21,598 INFO [decode.py:484] Device: cuda:0
2022-05-13 11:28:21,605 INFO [decode.py:493] {'best_train_loss': inf, 'best_valid_loss': inf, 'best_train_epoch': -1, 'best_valid_epoch': -1, 'batch_
idx_train': 0, 'log_interval': 500, 'reset_interval': 2000, 'valid_interval': 20000, 'feature_dim': 80, 'subsampling_factor': 4, 'encoder_dim': 512,
'nhead': 8, 'dim_feedforward': 2048, 'num_encoder_layers': 12, 'decoder_dim': 512, 'joiner_dim': 512, 'model_warm_step': 20000, 'env_info': {'k2-vers
ion': '1.15.1', 'k2-build-type': 'Release', 'k2-with-cuda': True, 'k2-git-sha1': 'f8d2dba06c000ffee36aab5b66f24e7c9809f116', 'k2-git-date': 'Thu Apr
21 12:20:34 2022', 'lhotse-version': '1.1.0.dev+missing.version.file', 'torch-version': '1.10.0+cu102', 'torch-cuda-available': True, 'torch-cuda-ver
sion': '10.2', 'python-version': '3.8', 'icefall-git-branch': 'master', 'icefall-git-sha1': '48a6a9a-clean', 'icefall-git-date': 'Fri May 13 11:03:26
 2022', 'icefall-path': '/ceph-fj/fangjun/open-source-2/icefall-master', 'k2-path': '/ceph-fj/fangjun/open-source-2/k2-multi-22/k2/python/k2/__init__
.py', 'lhotse-path': '/ceph-fj/fangjun/open-source-2/lhotse-master/lhotse/__init__.py', 'hostname': 'de-74279-k2-train-2-0307200233-b554c565c-lf9qd',
 'IP address': '10.177.74.201'}, 'epoch': 11, 'avg': 1, 'avg_last_n': 0, 'exp_dir': PosixPath('pruned_transducer_stateless2/exp'), 'bpe_model': 'data
/lang_bpe_500/bpe.model', 'decoding_method': 'greedy_search', 'beam_size': 4, 'beam': 4, 'max_contexts': 4, 'max_states': 8, 'context_size': 2, 'max_
sym_per_frame': 1, 'manifest_dir': PosixPath('data/fbank'), 'max_duration': 600, 'bucketing_sampler': True, 'num_buckets': 30, 'concatenate_cuts': Fa
lse, 'duration_factor': 1.0, 'gap': 1.0, 'on_the_fly_feats': False, 'shuffle': True, 'return_cuts': True, 'num_workers': 2, 'enable_spec_aug': True,
'spec_aug_time_warp_factor': 80, 'enable_musan': True, 'subset': 'XL', 'small_dev': False, 'res_dir': PosixPath('pruned_transducer_stateless2/exp/greedy_search'), 'suffix': 'epoch-11-avg-1-context-2-max-sym-per-frame-1', 'blank_id': 0, 'vocab_size': 500}
2022-05-13 11:28:21,605 INFO [decode.py:495] About to create model
2022-05-13 11:28:22,243 INFO [checkpoint.py:112] Loading checkpoint from pruned_transducer_stateless2/exp/epoch-11.pt
2022-05-13 11:28:31,272 INFO [decode.py:525] Number of model parameters: 78648040
2022-05-13 11:28:31,273 INFO [asr_datamodule.py:406] About to get dev cuts
2022-05-13 11:28:31,845 INFO [asr_datamodule.py:415] About to get test cuts
2022-05-13 11:28:35,545 INFO [decode.py:397] batch 0/?, cuts processed until now is 99
2022-05-13 11:29:24,008 INFO [decode.py:415] The transcripts are stored in pruned_transducer_stateless2/exp/greedy_search/recogs-dev-greedy_search-epoch-11-avg-1-context-2-max-sym-per-frame-1.txt
2022-05-13 11:29:24,198 INFO [utils.py:405] [dev-greedy_search] %WER 10.59% [13534 / 127790, 2987 ins, 3472 del, 7075 sub ]
2022-05-13 11:29:24,600 INFO [decode.py:428] Wrote detailed error stats to pruned_transducer_stateless2/exp/greedy_search/errs-dev-greedy_search-epoch-11-avg-1-context-2-max-sym-per-frame-1.txt
2022-05-13 11:29:24,601 INFO [decode.py:445]
For dev, WER of different settings are:
greedy_search   10.59   best for dev

2022-05-13 11:29:26,661 INFO [decode.py:397] batch 0/?, cuts processed until now is 118
2022-05-13 11:30:48,172 INFO [decode.py:397] batch 100/?, cuts processed until now is 10038
2022-05-13 11:31:41,922 INFO [decode.py:397] batch 200/?, cuts processed until now is 18926
2022-05-13 11:31:52,114 INFO [decode.py:415] The transcripts are stored in pruned_transducer_stateless2/exp/greedy_search/recogs-test-greedy_search-epoch-11-avg-1-context-2-max-sym-per-frame-1.txt
2022-05-13 11:31:52,676 INFO [utils.py:405] [test-greedy_search] %WER 10.87% [42460 / 390744, 6906 ins, 11246 del, 24308 sub ]
2022-05-13 11:31:54,014 INFO [decode.py:428] Wrote detailed error stats to pruned_transducer_stateless2/exp/greedy_search/errs-test-greedy_search-epoch-11-avg-1-context-2-max-sym-per-frame-1.txt
2022-05-13 11:31:54,015 INFO [decode.py:445]
For test, WER of different settings are:
greedy_search   10.87   best for test
```

One thing to note is that it takes only 1 minute to decode the `dev` dataset, while it takes about 1 hour and 18 minutes for conformer-ctc (from the [log](https://huggingface.co/wgb14/icefall-asr-gigaspeech-conformer-ctc/blob/main/log/log-decode-2022-04-08-22-02-12))

**Comment by wgb14** [comment]
> @wgb14
> 
> Could you also upload the decoding results, e.g., files like `log-*, errs-*, recogs-*`, to huggingface?

Sure, will do.

> 
> One thing to note is that it takes only 1 minute to decode the `dev` dataset, while it takes about 1 hour and 18 minutes for conformer-ctc (from the [log](https://huggingface.co/wgb14/icefall-asr-gigaspeech-conformer-ctc/blob/main/log/log-decode-2022-04-08-22-02-12))

It seems that your cluster has a faster IO or larger memory cache. I'm redoing the decoding with `--max-duration 600 --num_workers 4`, and it takes me about 15 minutes to decode dev set with `greedy-search`

```log
2022-05-13 11:52:35,758 INFO [decode.py:482] Decoding started
2022-05-13 11:52:35,759 INFO [decode.py:488] Device: cuda:0
2022-05-13 11:52:35,799 INFO [decode.py:497] {'best_train_loss': inf, 'best_valid_loss': inf, 'best_train_epoch': -1, 'best_valid_epoch': -1, 'batch_idx_train': 0, 'log_interval': 500, 'reset_interval': 2000, 'valid_interval': 20000, 'feature_dim': 80, 'subsampling_factor': 4, 'encoder_dim': 512, 'nhead': 8, 'dim_feedforward': 2048, 'num_encoder_layers': 12, 'decoder_dim': 512, 'joiner_dim': 512, 'model_warm_step': 20000, 'env_info': {'k2-version': '1.14', 'k2-build-type': 'Release', 'k2-with-cuda': True, 'k2-git-sha1': '1b29f0a946f50186aaa82df46a59f492ade9692b', 'k2-git-date': 'Wed Apr 13 08:46:49 2022', 'lhotse-version': '1.1.0', 'torch-version': '1.10.0', 'torch-cuda-available': True, 'torch-cuda-version': '11.1', 'python-version': '3.7', 'icefall-git-branch': 'gigaspeech_rnnt', 'icefall-git-sha1': '2d07df5-dirty', 'icefall-git-date': 'Sun Apr 17 10:01:21 2022', 'icefall-path': '/userhome/user/guanbo/icefall_rnnt', 'k2-path': '/opt/conda/lib/python3.7/site-packages/k2-1.14.dev20220513+cuda11.1.torch1.10.0-py3.7-linux-x86_64.egg/k2/__init__.py', 'lhotse-path': '/userhome/user/guanbo/lhotse/lhotse/__init__.py', 'hostname': 'c168bad00d26f011ec09c520b9eb77fc7b8a-chenx8564-0', 'IP address': '10.104.201.14'}, 'epoch': 29, 'avg': 11, 'avg_last_n': 0, 'exp_dir': PosixPath('pruned_transducer_stateless2/exp_500_8'), 'bpe_model': 'data/lang_bpe_500/bpe.model', 'decoding_method': 'greedy_search', 'beam_size': 4, 'beam': 4, 'max_contexts': 4, 'max_states': 8, 'context_size': 2, 'max_sym_per_frame': 1, 'manifest_dir': PosixPath('data/fbank'), 'max_duration': 600, 'bucketing_sampler': True, 'num_buckets': 30, 'concatenate_cuts': False, 'duration_factor': 1.0, 'gap': 1.0, 'on_the_fly_feats': False, 'shuffle': True, 'return_cuts': True, 'num_workers': 4, 'enable_spec_aug': True, 'spec_aug_time_warp_factor': 80, 'enable_musan': True, 'subset': 'XL', 'small_dev': False, 'res_dir': PosixPath('pruned_transducer_stateless2/exp_500_8/greedy_search'), 'suffix': 'epoch-29-avg-11-context-2-max-sym-per-frame-1', 'blank_id': 0, 'vocab_size': 500}
2022-05-13 11:52:35,800 INFO [decode.py:499] About to create model
2022-05-13 11:52:36,199 INFO [decode.py:515] averaging ['pruned_transducer_stateless2/exp_500_8/epoch-19.pt', 'pruned_transducer_stateless2/exp_500_8/epoch-20.pt', 'pruned_transducer_stateless2/exp_500_8/epoch-21.pt', 'pruned_transducer_stateless2/exp_500_8/epoch-22.pt', 'pruned_transducer_stateless2/exp_500_8/epoch-23.pt', 'pruned_transducer_stateless2/exp_500_8/epoch-24.pt', 'pruned_transducer_stateless2/exp_500_8/epoch-25.pt', 'pruned_transducer_stateless2/exp_500_8/epoch-26.pt', 'pruned_transducer_stateless2/exp_500_8/epoch-27.pt', 'pruned_transducer_stateless2/exp_500_8/epoch-28.pt', 'pruned_transducer_stateless2/exp_500_8/epoch-29.pt']
2022-05-13 11:55:48,958 INFO [decode.py:529] Number of model parameters: 78648040
2022-05-13 11:55:48,958 INFO [asr_datamodule.py:406] About to get dev cuts
2022-05-13 11:55:49,292 INFO [asr_datamodule.py:415] About to get test cuts
2022-05-13 11:56:45,876 INFO [decode.py:398] batch 0/?, cuts processed until now is 99
2022-05-13 12:10:29,422 INFO [decode.py:415] The transcripts are stored in pruned_transducer_stateless2/exp_500_8/greedy_search/recogs-dev-greedy_search-epoch-29-avg-11-context-2-max-sym-per-frame-1.txt
2022-05-13 12:10:29,589 INFO [utils.py:406] [dev-greedy_search] %WER 10.59% [13536 / 127790, 2978 ins, 3466 del, 7092 sub ]
2022-05-13 12:10:29,980 INFO [decode.py:428] Wrote detailed error stats to pruned_transducer_stateless2/exp_500_8/greedy_search/errs-dev-greedy_search-epoch-29-avg-11-context-2-max-sym-per-frame-1.txt
2022-05-13 12:10:30,011 INFO [decode.py:449] 
For dev, WER of different settings are:
greedy_search	10.59	best for dev
```

**Comment by csukuangfj** [comment]
> It seems that your cluster has a faster IO or larger memory cache.

I have created a google colab notebook to decode gigaspeech.
See https://colab.research.google.com/drive/14FL2q0uQt3hC4TQ61lseV3wM3gxbY58R?usp=sharing

It takes less than 3 minutes in the Colab notebook to decode the dev dataset with greedy search.

Screen Shot 2022-05-13 at 14 43 15

**Comment by xIaott-s** [comment]
> Thanks, Fangjun. These really help a lot. No, I'm still training with `--max-duration 120`, same with Conformer-CTC setup. I'm using torch 1.10.0, a version after that fix.

I had the same problem as you，and I'm also using torch 1.10.0. Only when I set --max-duration=120 can the training continue. 
Did you find the reason？Is this a bug for torch==1.10.0？

**Pull request merged**